src/pymoca/backends/casadi/model.py

```diff
@@ -500,17 +500,17 @@ def simplify(self, options):
 
             alg_states = OrderedDict([(s.symbol.name(), s) for s in self.alg_states])
 
-            variables = []
-            values = []
-
-            reduced_equations = []
-            for eq in self.equations:
+            def extract_assignment(eq):
                 if eq.is_symbolic() and eq.name() in alg_states and p.match(eq.name()):
-                    variables.append(eq)
-                    values.append(0.0)
-                    del alg_states[eq.name()]
-                    # Skip this equation
-                    continue
+                    return eq, 0.0
+
+                if eq.is_op(ca.OP_IF_ELSE_ZERO):
+                    cond = eq.dep(0)
+                    nested_eq = eq.dep(1)
+
+                    variable, value = extract_assignment(nested_eq)
+                    if variable is not None:
+                        return variable, ca.if_else(cond, value, 0, True)
```

Comment:
Why the short-circuiting?

Reply:
We short-circuit everywhere in generator.py. If we would not short-circuit here, we would change the short-circuiting behaviour of the processed if/else statements.

Reply:
How do we end up with a `ca.IF_ELSE_ZERO` then? 

```Python
x = ca.MX.sym("x")
y = ca.MX.sym("y")
cond = ca.MX.sym("cond")

eq = ca.if_else(cond, y - x, 0, True)
eq.is_op(ca.OP_IF_ELSE_ZERO)  # --> False

eq2 = ca.if_else(cond, y - x, 0)
eq2.is_op(ca.OP_IF_ELSE_ZERO)  # --> True
```

Reply:
```
eq = ca.if_else(cond, y, x, True)                                                                                                                     
[eq] = _expand_simplify_mx([eq])                                                                                                                      

eq.is_op(ca.OP_ADD)  # True                                                                                                                            
eq.dep(0).is_op(ca.OP_IF_ELSE_ZERO)  # True                                                                                                              
eq.dep(1).is_op(ca.OP_IF_ELSE_ZERO)  # True                                                                                                     
```

Reply:
Ah, so that's how.